contracts/instance/InstanceAccessManager.sol

```diff
@@ -1,292 +1,526 @@
     }
 
     function getRoleId(uint256 idx) external view returns (RoleId roleId) {
-        return _roles[idx];
+        return _roleIds[idx];
     }
 
+    // TODO now: for non existent name returns ADMIN_ROLE id
     function getRoleIdForName(string memory name) external view returns (RoleId roleId) {
-        return _roleForName[ShortStrings.toShortString(name)];
+        return _roleIdForName[ShortStrings.toShortString(name)];
     }
 
-    function getRole(RoleId roleId) external view returns (IAccess.RoleInfo memory role) {
-        return _role[roleId];
+    function roleMember(RoleId roleId, uint256 idx) external view returns (address member) {
+        return EnumerableSet.at(_roleMembers[roleId], idx);
     }
 
     function hasRole(RoleId roleId, address account) external view returns (bool accountHasRole) {
         (accountHasRole, ) = _accessManager.hasRole(roleId.toInt(), account);
     }
 
-    function roleMembers(RoleId roleId) external view returns (uint256 numberOfMembers) {
-        return EnumerableSet.length(_roleMembers[roleId]);
-    }
-
-    function getRoleMember(RoleId roleId, uint256 idx) external view returns (address roleMember) {
-        return EnumerableSet.at(_roleMembers[roleId], idx);
+    function roles() external view returns (uint256 numberOfRoles) {
+        return _roleIds.length;
     }
 
     //--- Target ------------------------------------------------------//
-    function createGifTarget(address target, string memory name) external restricted() {
-        _createTarget(target, name, false, true);
+    // ADMIN_ROLE
+    // assume some core targets are registred (instance) while others are not (instance accesss manager, instance reader, bundle manager)
+    function createCoreTarget(address target, string memory name) external restricted() {
+        _createTarget(target, name, IAccess.Type.Core);
     }
-
-    function createTarget(address target, string memory name) external restricted() {
-        _createTarget(target, name, true, true);
+    // INSTANCE_SERVICE_ROLE
+    // assume gif target is registered and belongs to the same instance as instance access manager
+    function createGifTarget(address target, string memory name) external restricted() 
+    {
+        _createTarget(target, name, IAccess.Type.Gif);
     }
+    // INSTANCE_OWNER_ROLE
+    // assume custom target.authority() is constant -> target can not be used with different instance access manager
+    // assume custom target can not be registered as component -> each service which is doing component registration MUST register a gif target
+    // assume custom target can not be registered as instance or service -> why?
+    // TODO check target associated with instance owner or instance or instance components or components helpers
+    function createTarget(address target, string memory name) 
+        external 
+        restricted() 
+    {
+        if(_registry.isRegistered(target)) {
```

Comment:
target do not have to be `Registrable`, they might be just simple contracts.

Reply:
In case instance owner wants to use registry, service, instance or component as custom target. While this registered target is not using the same authority it won't break anything but anyway will be meaningless, so better to prohibit such cases.

Reply:
deleted this check and checking for registration in createGifTarget()